### PR TITLE
Run taskRun without a Task by adding TaskSpec

### LINKED ISSUE #249: Execute a TaskRun without having to create a Task
Related to #147

## Expected Behavior

As a former Build user, I should be able to create a single resource that executes requested steps and can be watched for updates.

## Actual Behavior

If I want to run some steps, I first have to define a `Task` that specifies those steps, then invoke it with a `TaskRun` that references that `Task`.

## Additional Info

One possible solution would be to have the `TaskRun` type take either a `taskRef` or a `taskSpec` (but not both, or neither). When a `TaskRun` is requested with a `taskSpec`, it would perform the same steps as if a `Task` had been created with the given spec, then invoked by the `TaskRun`, without creating a `Task` as a side effect. Tasks defined in this way can't specify parameters, but can specify inputs. Resources might also need to be able to be defined without referencing a pre-existing CRD resource.

### PR DESCRIPTION
**Pull request by pivotal-nader-ziada**
Fixes #249 

## Proposed Changes

 * Add a **TaskSpec** field to the `TaskRun` which is _mutually exclusive_ with taskRef
 * validation for not having taskRef and taskSpec together but having one of them
 * Changed TaskRef field in taskRun to a pointer now that's optional
 * Unit tests for running `TaskRun` without a task
 ~~if TaskSpec includes `sources` it doesn't need to be defined as a `Resource`~~

now with this feature, users can run `TaskRun` without a task as demonstrated by the following example:

```yaml
apiVersion: pipeline.knative.dev/v1alpha1
kind: PipelineResource
metadata:
  name: go-example-git
spec:
  type: git
  params:
  - name: url
    value: https://github.com/pivotal-nader-ziada/gohelloworld
---
apiVersion: pipeline.knative.dev/v1alpha1
kind: TaskRun
metadata:
  name: build-push-task-run-2
spec:
  trigger:
    triggerRef:
      type: manual
  inputs:
    resources:
    - name: workspace
      resourceRef:
        name: go-example-git
  taskSpec: 
    inputs:
      resources:
      - name: workspace
        type: git    
    steps:
    - name: build-and-push
      image: gcr.io/kaniko-project/executor
      command:
      - /kaniko/executor
      args:
      - --destination=gcr.io/my-project/gohelloworld
```

### CODE REVIEW

File: pkg/reconciler/v1alpha1/taskrun/validate.go [hunk 1]
```diff
@@ -24,14 +24,17 @@ import (
...
+		// and all inputs/outputs are bound
+		t, err := c.taskLister.Tasks(tr.Namespace).Get(tr.Spec.TaskRef.Name)
+		if err != nil {
+			return fmt.Errorf("Error listing task ref %s: %v",
+				tr.Spec.TaskRef.Name, err)
+		}
+		return validateTaskRunAndTask(c, *tr, t, tr.Namespace)
 	}
-	return validateTaskRunAndTask(c, *tr, t, tr.Namespace)
+	return nil
```
Comment (vdemeester) [inline pkg/reconciler/v1alpha1/taskrun/validate.go]: Shouldn't we validate it in that case too ? :thinking: 

This would require to change the signature of  `validateTaskRunTask` though, but most likely for the good

```go
// […]
    return validateTaskRunAndTask(c, *tr, t.Spec, tr.Namespace)
}
return validateTaskRunAndTask(c, *tr, tr.TaskSpec, tr.Namespace)
``

(On that note, thinking outloud, we *only* validate the spec, so we may want to pass only the specs — the `tr.Name` is only required on the error message, so we could use `github.com/pkg/errors` to wrap errors and thus adding the `tr.Name` via the caller)

File: pkg/apis/pipeline/v1alpha1/taskrun_validation.go [hunk 2]
```diff
@@ -38,9 +38,15 @@ func (ts *TaskRunSpec) Validate() *apis.FieldError {
 		return apis.ErrMissingField("spec")
 	}
 
-	// Check for TaskRef
-	if ts.TaskRef.Name == "" {
-		return apis.ErrMissingField("spec.taskref.name")
+	// can't have both taskRef and taskSpec at the same time
+	if (ts.TaskRef != nil && ts.TaskRef.Name != "") && ts.TaskSpec != nil {
+		return apis.ErrDisallowedFields("spec.taskSpec")
```
Comment (shashwathi) [inline pkg/apis/pipeline/v1alpha1/taskrun_validation.go]: If this field takes multiple values then consider adding `spec.taskRef` along with `spec.taskSpec`

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go [hunk 3]
```diff
@@ -334,6 +334,33 @@ func TestReconcile(t *testing.T) {
...
+			Namespace: "foo",
+		},
+		Spec: v1alpha1.TaskRunSpec{
+			TaskSpec: &v1alpha1.TaskSpec{
+				Inputs: &v1alpha1.Inputs{
+					Params: []v1alpha1.TaskParam{{
+						Name:        "myarg",
+						Description: "mydesc",
+						Default:     "mydefault",
+					}},
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: looks like maybe we DO support params! i cant see a strong reason not support them esp. if we get it for free

File: pkg/reconciler/v1alpha1/taskrun/taskrun_test.go [hunk 4]
```diff
@@ -945,3 +972,207 @@ func TestCreateRedirectedBuild(t *testing.T) {
 		t.Errorf("services accounts do not match: %s should be %s", b.Spec.ServiceAccountName, tr.Spec.ServiceAccount)
 	}
 }
+
+func TestTaskRunWithTaskSpec(t *testing.T) {
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: probably dont need the extra case in test test above if we have this (or vice versa)

Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/taskrun/taskrun_test.go]: maybe prefer adding cases to the existing test so we dont need to duplicate the assertions?

File: pkg/apis/pipeline/v1alpha1/taskrun_validation.go (hunk 2 above)
Comment (pivotal-nader-ziada) [reply pkg/apis/pipeline/v1alpha1/taskrun_validation.go]: added

**Comment by pivotal-nader-ziada** [comment]
@bobcatfish addressed the comments, ready for another look

File: pkg/reconciler/v1alpha1/taskrun/resources/taskrunresolution.go [hunk 5]
```diff
@@ -70,3 +72,23 @@ func ResolveTaskRun(tr *v1alpha1.TaskRunSpec, getTask GetTask, gr GetResource) (
...
+		if err != nil {
+			return nil, taskName, fmt.Errorf("error when listing tasks for taskRun %s %v", taskRun.Name, err)
+		}
+		taskSpec = &t.Spec
+		taskName = t.Name
+	} else if taskRun.Spec.TaskSpec != nil {
+		taskSpec = taskRun.Spec.TaskSpec
+		taskName = taskRun.Name
+	} else {
+		return taskSpec, taskName, fmt.Errorf("TaskRun %s not providing TaskRef or TaskSpec", taskRun.Name)
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/resources/taskrunresolution.go]: this happens to be the only line not covered by the unit tests and i actually wonder if we need it at all - at this point, we do happen to know that the taskSpec has been validated, and there should be no case like this

however maybe its safer to keep it - im fine with not covering the line in unit tests :D

**Comment by bobcatfish** [comment]
Looks great to me! Thanks for adding the docs also ^.^

/lgtm

I'm going to add a hold here b/c @ImJasonH opened the issue and might have some opinions about us removing `sources`

/hold

/meow space

File: docs/using.md [hunk 6]
```diff
@@ -183,6 +183,46 @@ See [the example PipelineRun](../examples/invocations/kritis-pipeline-run.yaml).
    that the `Task` needs to run.
 2. The `TaskRun` will also serve as a record of the history of the invocations of the
    `Task`.
+3. Another way of running a Task is embedding the TaskSpec in the taskRun yaml as shown in the following example
+
+```yaml
+apiVersion: pipeline.knative.dev/v1alpha1
```
Comment (ImJasonH) [inline docs/using.md]: So we still need to create two resources to run a simple set of steps? How can we drop the requirement that a PipelineResource already exists?

In Build, resources were specified directly in the build itself, as `source(s)`.

File: docs/using.md [hunk 7]
```diff
@@ -183,6 +183,46 @@ See [the example PipelineRun](../examples/invocations/kritis-pipeline-run.yaml).
...
+  inputs:
+    resources:
+    - name: workspace
+      resourceRef:
+        name: go-example-git
+  taskSpec: 
+    inputs:
+      resources:
+      - name: workspace
+        type: git
```
Comment (ImJasonH) [inline docs/using.md]: This type is specified twice, once in the type of the `workspace` resource (line 194) and again here. Can one be removed? What happens if the resource is `type:foo` but I specify it as an input with `type:bar` ?

File: pkg/apis/pipeline/v1alpha1/taskrun_validation.go [hunk 8]
```diff
@@ -38,9 +38,16 @@ func (ts *TaskRunSpec) Validate() *apis.FieldError {
 		return apis.ErrMissingField("spec")
 	}
 
-	// Check for TaskRef
-	if ts.TaskRef.Name == "" {
-		return apis.ErrMissingField("spec.taskref.name")
+	// can't have both taskRef and taskSpec at the same time
+	if (ts.TaskRef != nil && ts.TaskRef.Name != "") && ts.TaskSpec != nil {
+		fields := []string{"spec.taskref", "spec.taskspec"}
+		return apis.ErrDisallowedFields(fields...)
```
Comment (ImJasonH) [inline pkg/apis/pipeline/v1alpha1/taskrun_validation.go]: This should work: 

`return apis.ErrDisallowedFields("spec.taskref", "spec.taskspec")`

File: pkg/apis/pipeline/v1alpha1/taskrun_validation.go [hunk 9]
```diff
@@ -38,9 +38,16 @@ func (ts *TaskRunSpec) Validate() *apis.FieldError {
...
+	}
+
+	// Check that one of TaskRef and TaskSpec is present
+	if (ts.TaskRef == nil || (ts.TaskRef != nil && ts.TaskRef.Name == "")) && ts.TaskSpec == nil {
+		fields := []string{"spec.taskref.name", "spec.taskspec"}
```
Comment (ImJasonH) [inline pkg/apis/pipeline/v1alpha1/taskrun_validation.go]: Here too

File: docs/using.md (hunk 6 above)
Comment (pivotal-nader-ziada) [reply docs/using.md]: I think adding even more embedding is just making the yaml file too complicated. Also one side benefit is that the resources are such a core concept in build-pipeline that we want users to know about when migrating.

Maybe as part of the work for the resources extensibility we can make this easier by making the taskRun controller create the resource for for system supported resource types such as git. In that case, the we can define types for system resources with their specific fields to simplify yaml files for all tasks

Depending on what we decide here, how big the change is, it might be better to have another follow up PR so this one doesn't get too big

**Comment by pivotal-nader-ziada** [comment]
/hold cancel

since I made a small fix and rebased which removed the lgtm

File: docs/using.md (hunk 6 above)
Comment (bobcatfish) [reply docs/using.md]: > I think adding even more embedding is just making the yaml file too complicated. Also one side benefit is that the resources are such a core concept in build-pipeline that we want users to know about when migrating.

That's a good point!

> Depending on what we decide here, how big the change is, it might be better to have another follow up PR so this one doesn't get too big

Sounds good to me

**Comment by ImJasonH** [comment]
/lgtm

Sounds like there are some more streaming changes we should make, but those definitely don't belong in this PR. 

Thanks for doing this work Nader, this should definitely improve the basic getting-started experience, and will make Serving usage simpler, once we get there.
